com.teamscale.polarion.plugin/src/com/teamscale/polarion/plugin/WorkItemUpdatesServlet.java

```diff
@@ -0,0 +1,769 @@
+   * generate a change/field diff for the opposite end of the link. So, we need to generate those
+   * link changes manually for receiving end.
+   *
+   * @param workItemId represents the link origin. On the map, this id will be the receiver
+   *     (reverse)
+   * @param revision represents the revision number when the link changed (added/removed)
+   * @param links represents a collection generated by Polarion containing the added/removed links
+   * @param added is true if this a list of added links, otherwise these are removed links *
+   */
+  private void updateOppositeLinksMap(
+      String workItemId, String revision, Collection<ILinkedWorkItemStruct> links, boolean added) {
+
+    // For each link struct, get the WI id, check if there's an entry in the map
+    // if there is, check if there's a link of same type, action (added/removed), and revision
+    // if there isn't, add an entry to the map flipping the ids (reverse)
+    for (ILinkedWorkItemStruct linkStruct : links) {
+      List<LinkBundle> linkBundles = backwardLinksTobeAdded.get(linkStruct.getLinkedItem().getId());
+      LinkBundle reverse = null;
+      if (linkBundles == null) {
+        reverse = new LinkBundle();
+        reverse.setAdded(added);
+        reverse.setRevision(revision);
+        reverse.setLinkedWorkItem(new LinkedWorkItem(workItemId, linkStruct.getLinkRole().getId()));
+        List<LinkBundle> newLinkBundles = new ArrayList<LinkBundle>();
+        newLinkBundles.add(reverse);
+        backwardLinksTobeAdded.put(linkStruct.getLinkedItem().getId(), newLinkBundles);
+      } else {
+        if (!alreadyHasLinkBundle(linkBundles, linkStruct, revision, added)) {
+          reverse = new LinkBundle();
+          reverse.setAdded(added);
+          reverse.setRevision(revision);
+          reverse.setLinkedWorkItem(
+              new LinkedWorkItem(workItemId, linkStruct.getLinkRole().getId()));
+          linkBundles.add(reverse);
+        }
+      }
+    }
+  }
+
+  /**
+   * Helper method called by {@link #updateOppositeLinksMap()} to check if a LinkBundle was already
+   * created in the map *
+   */
+  private boolean alreadyHasLinkBundle(
+      List<LinkBundle> linkBundles,
+      ILinkedWorkItemStruct linkStruct,
+      String revision,
+      boolean added) {
+    if (linkBundles == null) return false;
+    for (LinkBundle linkBundle : linkBundles) {
+      if (linkBundle.getRevision().equals(revision)
+          && linkBundle.isAdded() == added
+          && linkBundle.getLinkedWorkItem().getLinkRoleId().equals(linkStruct.getLinkRole().getId())
+          && linkBundle.getLinkedWorkItem().getId().equals(linkStruct.getLinkedItem().getId())) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /** Binary search to cut down the search space since the list is ordered in ascending order* */
+  private int searchIndexWorkItemHistory(IPObjectList<IWorkItem> workItemHistory) {
+
+    if (workItemHistory == null) return -1;
```

Comment:
Should never happen as this was checked beforehand. Throw NPE instead

Reply:
Done